lib/std/strbasics.nim

```diff
@@ -27,6 +27,21 @@ proc add*(x: var string, y: openArray[char]) =
     i.inc
   # xxx use `nimCopyMem(x[n].addr, y[0].addr, y.len)` after some refactoring
 
+proc strnlen(s: cstring, maxlen: csize_t): csize_t {.importc, header: "<string.h>".}
+
+func addCstring*(result: var string, buf: openArray[char]) =
+  ## Appends to `result` chars from `buf` up to (and excluding) the 1st ``\0``, if any.
+  runnableExamples:
+    var a = "foo"
+    a.addCstring(['a', 'b', '\0', 'd'])
+    assert a == "fooab"
+    a.addCstring "123"
+    a.add "45".cstring # for cstring, use `add` directly
+    assert a == "fooab12345"
+  if buf.len > 0:
+    let n = strnlen(cast[cstring](buf[0].unsafeAddr), buf.len.csize_t).int
+    add(result, toOpenArray(buf, 0, n - 1))
```

Comment:
Why not use a loop or copyMem?

Reply:
> copyMem

copyMem would not help (it doesn't tell you where the 1st \0 is) so would still require strnlen

> Why not use a loop 

because strnlen is optimized, see this benchmark I just wrote showing 7x speedup: https://github.com/timotheecour/Nim/issues/751
(it also has some links explaining some of the tricks strnlen uses, a lot of which could be reusable in nim string utilities)

Reply:
The loops don't prune type information though and simply work for the Nim VM and JS. If too slow, the C codegen should recognize some patterns and generate `copyMem` and the like. Easier to maintain than the `when` jungle. We could create a pragma like `.mustOptimize.` to ensure no performance regressions (the codegen would bail out if it failed to detect the pattern even though the loop was annotated with `.mustOptimize`).

Reply:
the "when jungle" to support this in nimvm/js doesn't add any complexity to client code and that's the only thing that matters. strbasics is a low-level module which offers zero-cost (or low-cost) abstractions hidden behind a clean API.